### ISSUE TITLE
AutoKey must be restarted to recognize changes to internal module

### ISSUE THREAD

**Issue by Elliria** [autokey/autokey#747]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Bug

### Which Linux distribution did you use?

Kubuntu 22.04 LTS

### Which AutoKey GUI did you use?

Both

### Which AutoKey version did you use?

0.95.10

### How did you install AutoKey?

My distribution's repository.

### Can you briefly describe the issue?

If you make a change to an internall module file once AutoKey has already interacted with it, AutoKey won't see the change until AutoKey is closed and restarted, which can cause you to believe you made a mistake in your newly-added or edited code when it may actually be that AutoKey just hasn't noticed your change yet.

### Can the issue be reproduced?

Always

### What are the steps to reproduce the issue?

1. Go into AutoKey's settings and choose a directory for AutoKey to use for internal modules.
2. Create the **MyModule.py** file in that modules directory with these contents:
```Python
#!/usr/bin/env python3

# Create a variable with a string as its value:
MyVariable="one"

# Create a function that returns a string:
def MyFunction():
	return "two"
```
3. Create an AutoKey script with these contents:
```Python
# Import the necessary library:
import MyModule

# Display the value of a variable from the imported module in a dialog:
dialog.info_dialog(message=MyModule.MyVariable)

# Display the value of a function from the imported module in a dialog:
dialog.info_dialog(message=MyModule.MyFunction())
```
4. Run the AutoKey script.
5. Observe the result. You will see a dialog with "**one**" in it followed by a dialog with "**two**" in it.
6. Edit the **MyModule.py** file to change "**two**" to "**three**" in the last line of the script.
7. Save the change to the **MyModule.py** file.
8. Run the AutoKey script.
9. Observe the result. You will see a dialog with "**one**" in it followed by a dialog with "**two**" in it.

### What should have happened?

Since you changed "**two**" to "**three**" in steps 6 and 7 above, you should have seen a dialog with "**one**" in it followed by a dialog with "**three**" in it when you observed the changes in step 9 above.

### What actually happened?

You saw a dialog with "**one**" in it followed by a dialog with "**two**" in it in step 9 above.

### Work-around:
Closing and restarting AutoKey (possibly also using `pkill autokey` before restarting AutoKey to make sure the process has ended) and running the AutoKey script again will cause AutoKey to properly display the current values of the variables in the internal module.

### Note:
This only happens when importing a module that's in the directory that you have chosen for AutoKey to use for modules. If you use [the steps I shared to run an imported external module from any directory](https://github.com/autokey/autokey/discussions/743) and make a change to the external script, AutoKey will instantly recognize it.

**Comment by josephj11** [comment]
In Linux, two immediate possibilities occur to me:

AutoKey/Python is somehow caching the Modules file, so the revisions are not seen.

AutoKey/Python has opened the file in Modules, but has not subsequently closed it. This would keep the original file attached to its file handle, etc. until it is explicitly closed or is closed as part of the AutoKey process itself closing when terminated.

When you import a module, it makes sense that Python would preserve it somehow to make the next import/access more efficient/faster.

**Comment by Elliria** [comment]
It's probably not Python, because it doesn't happen when using [the steps I shared to run an imported external module from any directory](https://github.com/autokey/autokey/discussions/743), which also imports a module, but recognizes any changes to it instantly. It could be, though, because that example does use a round-about way of doing the importing, so it may avoid whatever it is that's happening when imports are done directly from AutoKey's Modules directory.

I suspect, though, that you're most likely right in the theory you mentioned in [an AutoKey discussion](https://github.com/autokey/autokey/discussions/743#discussioncomment-4030503) in which you suggested that AutoKey has a list of directories it watches for changes and that if the Modules directory isn't on that list, that it may need to be added.

**Comment by ineuw** [comment]
Do you consider this an issue? I always took it granted. :-)

**Comment by josephj11** [comment]
@ineuw I haven't seen anyone actually use the Modules folder at all (mainly because it's a PITA to get API calls to work in functions in modules).

Do you use this feature at all?

A while back, AutoKey was enhanced to watch its directories and reload things when scripts or phrases are changed externally. It would be consistent for the Modules directory to be  watched as well.

**Comment by Elliria** [comment]
Since it only happens when you make a change to a module that was imported from the directory that you've chosen for AutoKey to use for modules and doesn't happen when you make a change to an external module, I'd consider that a discrepancy in behavior that should be made consistent.

I use it only for test purposes, but there could be countless numbers of others who aren't participating in this issue report who use it and would benefit from the fix. I'll be looking at the code for it by and by.

**Comment by ineuw** [comment]
To detail my experience, any change to Autokey made through its GUI interface, I close and restart Autokey. For me, it's a rule because a change in a script is not picked up otherwise. I do the same when renaming scripts.

**Never** edit script files in an external text editor, unless Autokey is closed. For this reason I keep the start icon in the panel next to the indicator.

Autokey must not run while using Wine, which I must use. All wine .desktop files use a script to kill Autokey before loading. The problem is not Autokey, but how X-Windows interfaces with Wine with the keyboard and mouse.

I must follow these rules not to have issues. It took me a long time to identify these, but it cleared a bit the haze of my ignorance.

**Comment by josephj11** [comment]
Thanks for this detailed explanation.

This seems a bit extreme. Notes inline.

Joe

> To detail my experience, any change to Autokey made through its GUI interface, I close and restart Autokey. For me, it's a rule because a change in a script is not picked up otherwise. I do the same when renaming scripts.
> 

I seem to recall some weirdness with externally renaming AutoKey actions, but it has never been documented. If anything doesn't work as expected, issues should be flled.

Any changes made in the GUI should take effect as soon as you select the Save button on the Main menu.

If you have enabled the AutoKey option to save automatically, the script will also be saved when you close the AutoKey window even if you haven't selected Save manually.

All editing changes appear to be done in memory or in a temporary file while the script is open in the editing panel. They are not saved until Save is selected or the AutoKey window is closed with autosaving enabled.

That's all that should be required. Anything that doesn't work this way is a bug and should be reported.

> **Never** edit script files in an external text editor, unless Autokey is closed. For this reason I keep the start icon in the panel next to the indicator.
> 
I do this all the time with no problems. AutoKey actively watches its directories and updates itself when changes are detected. The only problem I have seen is that the script currently displayed in the editing panel doesn't get updated. This can be remedied by switching to another script and then switching back or by closing the AutoKey main menu window and reopening it.

If you edit a script externally while that same script is displayed in the edit panel and then either manually or automatically save changes to that script within AutoKey, the new, externally edited version of the script might be overwritten by the version from the editing panel. If this occurs, it's also a bug.

AutoKey should save the modification time of the script when it is first loaded into the panel. If AutoKey tries to subsequently save the script, it could compare the script's current last modified time with what it was when it was first loaded into the editing panel. If these are not identical, then the script was modified externally after it was loaded into AutoKey and it should at least ask you what to do.

AutoKey does not maintain any database for actions. They are all stored exclusively in their respective files and the associated metadata sidecar files. It does not do any caching either.

It should never be necessary to terminate AutoKey when external changes are made.

So, anything that does not work as above is a bug and should be reported.

> Autokey must not run while using Wine, which I must use. All wine .desktop files use a script to kill Autokey before loading. The problem is not Autokey, but how X-Windows interfaces with Wine with the keyboard and mouse.
> 
I don't use wine. It has almost never worked for any app I cared about. But I haven't tried it in years. One exception to this is RegexBuddy 4 which I got somewhere. It works fine and can be quite helpful with developing and testing new regexes.

I would think that simply disabling AutoKey expansions should be enough for wine to work correctly. If it's not (and I'm sure you must have tried that), it's probably a bug in wine and/or AutoKey and I _really_ don't want to go down that rabbit hole, but it would be good to file one or more issues about it just so others can find them if they run into similar problems. Maybe something should also be added to the Known Limitations wiki page.

> I must follow these rules not to have issues. It took me a long time to identify these, but it cleared a bit the haze of my ignorance.

**Comment by ineuw** [comment]
I have no idea what expansions are. :-) I see it on the menu, but never asked. :-)

**Comment by josephj11** [comment]
Turning it off stops AutoKey  from responding to any trigger abbreviations or hotkeys. Enabling it again makes AutoKey work normally. It's called Enable monitoring. I got the wording wrong above.

"Expansions" is a term for expanding an abbreviation into its replacement text. Monitoring is better because expanding isn't all that AutoKey does.

**Comment by Elliria** [comment]
Come to think of it, I also sometimes have to restart AutoKey when renaming a script, but not always. That's worthy of an issue report. There wasn't an existing one, so issue #884 has now been created.

If Wine is managing to affect unrelated other programs that happen to be open at the same time as it is, that's a pretty serious issue that should probably be reported to the Xorg and Wine teams and/or to the team in charge of the desktop and/or window manager that you use.

For me, that would be a good reason not to use Wine. I used Wine a few years ago, but it was so hard getting certain games to play nice with it that I finally decided to stop using it. If you can, you might want to do what I did:

I bought a mechanical switch that lives in a DVD bay in my machine and physically disconnects and connects my drives at the push of its buttons, bought a new drive, connected my drives to the mechanical switch, installed Windows onto the new drive, and can now press buttons on the mechanical switch before booting to connect the drive I want to boot with and disconnect the drives I don't want to boot with. That way, I can run Windows programs on that drive without that drive even being aware that my other drives exist (and vice versa).

**Comment by ineuw** [comment]
Wine software works beautifully within themselves. Autokey interferes with the keyboard because the apps have extensive keyboard macros. Quitting Autokey automatically before starting an app works well for me since I work only in one window. 

This is a sample shell file

```
#!/usr/bin/bash
# 2023-07-18 06:33 - app_akgtk_is_dead_textpad.sh
pkill -f 'autokey-gtk'

# start textpad
env WINEPREFIX="/home/ineuw/.wine/dosdevices/h:/.wine" wine "/home/ineuw/.wine/drive_c/Program Files/TextPad 8/TextPad.exe"
```

**Comment by josephj11** [comment]
That works.

I would make it more generic by using `pkill autokey`. Then it will also work if you happen to be running autokey-qt for testing, etc.

What problem did you run into that made you add the `-f` (full match) option?

**Comment by ineuw** [comment]
I usually find the best discussions are on StackOverflow and StackExchange. There were a half a dozen different solutions for different scenarios. `pkill -f` was the best option after reading up on the pkill from the outset. I didn't have a problem.

P.S: I tested it with autokey-qt as well. There is no difference.

**Comment by ineuw** [comment]
@josephj11, about Enable Expansions. I tried it but it makes no difference when unchecked, I also find it re-checked after awhile.

**Comment by Elliria** [comment]
You might want to jump in on issue #873, @ineuw. The same thing happens to me.

**Comment by Elliria** [comment]
>Wine software works beautifully within themselves. Autokey interferes with the keyboard because the apps have extensive keyboard macros. Quitting Autokey automatically before starting an app works well for me since I work only in one window.

Thanks for the explanation that nicely lets AutoKey off the hook, which is always a good thing.

As for Wine software working beautifully, I guess it depends on the software. Many of the games we ran in it (RPGs, MMORPGs, and his first-person-shooters) took an incredible amount of wrestling to put into place, and once there, they often had annoying and sometimes completely debilitating issues. Very rarely did anything run perfectly. Even more frustrating would be a trip to the [WineHQ App Database](https://appdb.winehq.org/), where you could find lots of, "Oh, all you have to do is [insert instructions here]," which always became another time-sink that usually didn't work for us.

In the end, I weighed the difference between lightening my wallet by buying a new drive, a mechanical drive switch, and Windows against not doing so and then spending all that time and aggravation every time we want to install something new to not even be able to reliably count on succeeding. The lightened wallet won.

**Comment by josephj11** [comment]
>@josephj11, about Enable Expansions. I tried it but it makes no difference when unchecked, I also find it re-checked after awhile.

If you can reproduce it getting reenabled by itself, that's a bug.

**Comment by ineuw** [comment]
Since I mentioned it, I am monitoring it. Should it remain unchecked after quitting and restart?

**Comment by josephj11** [comment]
IDK. but it seems counterintuitive to me. Why would I launch AutoKey if I didn't want it to be active - at least to start with?

**Comment by ineuw** [comment]
After opening this topic, I kept on testing, and it turns on after saving something.

How can I keep it off, and use wine without having to kill AK?.

I also turn it off because of old habits in both Linux and Windows. A restart is preferable to wasting time on guessing.

We cannot say that a software is working 100% in all environments, when it depends on numerous other software sources and the changes and updates which we don`t know about until it affects us.

**Comment by josephj11** [comment]
Monitoring turning itself on after a save appears to be a bug and needs a separate issue.

----

We don't claim 100% of anything except perhaps enthusiam for AutoKey. Count the number of open issues to verify this.

Nothing is ever 100% by that definition. Computers and software almost always have unintended interactions at some point and if they don't now, future changes in the "environment" may create them.
We just have to do the the best we can with what we already know about and keep testing over time to find new problems.

**Comment by Elliria** [comment]
Monitoring is enabled for me on start after disabling it and shutting down AutoKey 0.95.10 and AutoKey 0.96.0 in both **autokey-gtk** and **autokey-qt** under Kubuntu 22.04 LTS. 

Interestingly, although both have a toggle for that setting in the tray icon's context menu, only **autokey-qt** also has a toggle for it in the **Settings** menu. Either way, it's fully enabled on restart. It seems like it should be an issue, so I filed it as issue #889.

**Comment by josephj11** [comment]
There are a few conceptually separate issues here.

One is that AutoKey appears to reenable monitoring by itself while it is still running if a Save operation is selected. That's clearly a bug.

It would be good to open a separate enhancement issue to add a Disable monitoring option to the GTK front end menu.

Another is that AutoKey always starts out with monitoring enabled when it is initially run. To me, this seems like the most reasonable thing to do.

If I don't want AutoKey active, why would I run it? Using external tools like editors or pagers, I can examine or change any action without invoking AutoKey. The only thing I can't do is have it actively working.

Manually changing a few parameters such as the `Trigger on` option  are a bit tricky to do correctly, but it's not something that anyone would normally do except when testing bugs. Generally, if I change anything, the first thing I want to do is see if it works and AutoKey has to be running for that.

The script editing panel does have some nice autocompletion features, but the UX is so bad that I almost never use it.

If I wanted that badly (which I don't), I would add that into kate. (Kate supports things like this if I could figure out how to do it.) If I got that working, I would open an enhancement issue to either have AutoKey launch and interact with a kate window or, more ambitiously, to replace qscintilla with kate-part (the guts of kate) in the Qt front end.

The editing panel is pretty small and popping out a separate window (that could be as large as full screen and repositioned as desired on the same or on a different virtual desktop) would provide a much better programming experience. This would be really great on systems with more than one display.

There's probably a way to do the same thing in the GTK environment.

I really like that idea and would definitely open an enhancement issue for it if we had sufficient developer resources so that it might actually get implemented.

**Comment by Elliria** [comment]
> One is that AutoKey appears to reenable monitoring by itself while it is still running if a Save operation is selected. That's clearly a bug.

Yes.
 
> It would be good to open a separate enhancement issue to add a Disable monitoring option to the GTK front end menu.

I agree. That would be consistent with **autokey-qt**.

It would also be a good idea to make the wording of the option the same on both **autokey_gtk** and **autokey_qt**. For example, choosing either **Enable Expansions** from **autokey-gtk** or **Enable Monitoring** from **autokey-qt** and using the same one on both.

> If I don't want AutoKey active, why would I run it?

You wouldn't, but this isn't about you. This is about everybody. There may be someone who would like AutoKey to be sitting ready and waiting in the tray, but not enabled.

> I really like that idea and would definitely open an enhancement issue for it if we had sufficient developer resources so that it might actually get implemented.

You should probably go for it while you're thinking of it and it can languish in the issues until a future developer snaps it up.

**Comment by ineuw** [comment]
@josephj11, your reasoning is faulty. Why was it included in the first place? I am sure that Deckter foresaw possible issues. It was you who suggested that it might be the source of my problem between wine and Autokey.

As for using wine, there is nothing that comes close to Textpad and Irfanview for versatility and ease of use.

**Comment by josephj11** [comment]
@Elliria No, it's not about me. I'm expressing my opinion, but I try to always be open to being convinced that there's a better way than whatever I have thought of and, if we can get several "votes" for something, I would almost always go with the majority opinion if we can get one. I would have to think something is really bad not to go along with it.

`Enable/Disable monitoring` is much more accurate IMO.

@ineuw I never suggested getting rid of the option or making it less available.  That it changes by itself within one session is definitely a bug that should be fixed. Having monitoring enabled by default on startup makes sense to me, but if others feel differently, that's fine and it can be changed.

If we do go with AutoKey remembering its monitoring state between executions, I think we should add a monitoring status indicator to the main window so people  can imediately see why nothing is working. We might also want to turn the panel icon gray and have a tooltip saying it's disabled.

**Comment by Elliria** [comment]
I like the idea of the icon changing and having a tooltip. The tooltip could also perhaps contain brief instructions for enabling AutoKey.

**Comment by Elliria** [comment]
Have you tried the Geany text editor, @ineuw? I wouldn't be without it.

**Comment by josephj11** [comment]
I have installed geany and code, but never got around to trying either of them.

I played around with idle a bit, but not enough to have an opinon on it. IDEs are nice, but since they don't support our API (AFAIK), they are of limited utility.

It would be great if you could set AutoKey to call out to your favorite editor instead of using the low-powered ones in the panel.

**Comment by Elliria** [comment]
Geany is an amazing text editor. It's basically Kate on steroids and it can do everything except make dinner for you.

I don't mind using the AutoKey editor for writing scripts, but if there's anything Pythonic that doesn't involve the AutoKey API and that needs thorough testing, I sometimes open up the empty Python script that lives permanently on my desktop and slap the code into there to work with and test from within Geany before putting it back into the AutoKey editor.
